cluster-autoscaler/proposals/parallel_drain.md

```diff
@@ -0,0 +1,246 @@
+    *   move nodes from candidate\_set to deleted\_set
+
+### Motivation for using target\_node\_pointer in draining simulation
+
+At first sight it may seem more natural to do the draining simulation in a way
+when we try to find the destination node starting from the tail of the sorted
+nodes list for each pod. That approach would probably result in tighter packing
+and a larger candidate set. Yet it comes at huge scalability cost:
+
+*   Much more simulation steps would need to be made as the nodes in the end of
+    the list are tighter packed already. Therefore there is a high chance that
+    we would need to traverse a large part of the list to find a destination
+    candidate. The situation would get worse as the algorithm progresses and
+    more rescheduling simulations are made.
+*   With that approach more nodes would be used as destinations for pods coming
+    from a single source node. As a result it would be less probable that we
+    could use the pod\_destination and pod\_sources maps effectively in the
+    "quick path"
+
+Using target\_node\_pointer approach mitigates above issues at a cost of
+(hopefully) not much worse simulation outcome.
+
+Implementation-wise, we'd like to reuse existing SchedulerBasedPredicateChecker.
+In order to do this, the algorithm for picking nodes for scheduling will be
+modified: it will be passed to SchedulerBasedPredicateChecker as a strategy. By
+default, it will be the existing round robin across all nodes, but ScaleDown
+simulation will inject a more sophisticated one, relying on the
+target\_node\_pointer.
+
+In the initial version, vanilla SchedulerBasedPredicateChecker will be used.
+Implementing support for target\_node\_pointer is essentially an optimization
+on top of it.
+
+### PDBs checking
+
+Throughout the loop we will keep the quotas remaining for PDBs in
+pdbs\_remaining\_disruptions structure. The structure will be computed at the
+beginning of UpdateUnneededNodes and for each PDB it will hold how many Pods
+matching this PDB can still be disrupted. Then we decrease the counters as we go
+over the nodes and simulate pods rescheduling. The initial computation and drain
+simulation takes into account the state of the pod. Specifically if Pod is not
+Healthy it will be subtracted from the remaining quota on the initial
+computation and then not subtracted again on drain simulation.
+
+
+### PreFilteringScaleDownNodeProcessor changes
+
+We need to repeat checks from PreFilteringScaleDownNodeProcessor in
+UpdateUnneededNodes anyway as the latter simulate multi node deletion so it
+seems we can drop PreFilteringScaleDownNodeProcessor if new scale down algorithm
+is enabled. Not crucial as checks are not costly.
+
+
+### Changes to ScaleDownStatus
+
+ScaleDownStatus does not play very well with the concept of multiple nodes being
+scheduled down. The existing ScaleDownInProgress value in ScaleDownResult enum
```

Comment:
> ScaleDownInProgress - the scale down wasn't attempted, because a previous scale-down was still in progress.

I think technically this is going to still be correct, but maybe `ScaleDownMaxParallelismReached` would be more accurate. @towca @MaciekPytel - any thoughts?

Reply:
I'm ok either way - I think the condition is analogous to the current one, especially if the current condition can show up if scale-down of empty nodes is rate limited (not sure if it can?). But `ScaleDownMaxParallelismReached` is indeed much more clear and I see no reason not to add it.

Reply:
Actually, empty node deletions can currently pile up - if they take a lot of time, there will be no error and CA will keep on scheduling more node deletions. That probably should result in some scale down error though.

Rate limiting is a slightly different mechanism than the one I'm proposing here. To avoid naming the status after implementation I decided to use `ScaleDownThrottled` instead, so we're free to change the throttling mechanism in the future without affecting the status type.